Put Unix sockets in /var/run/xrdp.

Closes: #656210
diff --git a/common/os_calls.c b/common/os_calls.c
index ceb581d..65d695a 100644
--- a/common/os_calls.c
+++ b/common/os_calls.c
@@ -664,16 +664,16 @@ g_create_wait_obj(char* name)
   if ((name == 0) || (strlen(name) == 0))
   {
     g_random((char*)&i, sizeof(i));
-    sprintf(sa.sun_path, "/tmp/auto%8.8x", i);
+    sprintf(sa.sun_path, "/var/run/xrdp/auto%8.8x", i);
     while (g_file_exist(sa.sun_path))
     {
       g_random((char*)&i, sizeof(i));
-      sprintf(sa.sun_path, "/tmp/auto%8.8x", i);
+      sprintf(sa.sun_path, "/var/run/xrdp/auto%8.8x", i);
     }
   }
   else
   {
-    sprintf(sa.sun_path, "/tmp/%s", name);
+    sprintf(sa.sun_path, "/var/run/xrdp/%s", name);
   }
   unlink(sa.sun_path);
   len = sizeof(sa);
diff --git a/sesman/chansrv/chansrv.c b/sesman/chansrv/chansrv.c
index ed1400f..bd9c426 100644
--- a/sesman/chansrv/chansrv.c
+++ b/sesman/chansrv/chansrv.c
@@ -409,7 +409,7 @@ setup_listen(void)
   if (g_use_unix_socket)
   {
     g_lis_trans = trans_create(2, 8192, 8192);
-    g_snprintf(port, 255, "/tmp/xrdp_chansrv_socket_%d", 7200 + g_display_num);
+    g_snprintf(port, 255, "/var/run/xrdp/xrdp_chansrv_socket_%d", 7200 + g_display_num);
   }
   else
   {
diff --git a/sesman/sessvc/sessvc.c b/sesman/sessvc/sessvc.c
index be84f9d..28d0183 100644
--- a/sesman/sessvc/sessvc.c
+++ b/sesman/sessvc/sessvc.c
@@ -56,12 +56,12 @@ chansrv_cleanup(int pid)
 {
   char text[256];
 
-  g_snprintf(text, 255, "/tmp/xrdp_chansrv_%8.8x_main_term", pid);
+  g_snprintf(text, 255, "/var/run/xrdp/xrdp_chansrv_%8.8x_main_term", pid);
   if (g_file_exist(text))
   {
     g_file_delete(text);
   }
-  g_snprintf(text, 255, "/tmp/xrdp_chansrv_%8.8x_thread_done", pid);
+  g_snprintf(text, 255, "/var/run/xrdp/xrdp_chansrv_%8.8x_thread_done", pid);
   if (g_file_exist(text))
   {
     g_file_delete(text);
diff --git a/xrdp/xrdp_mm.c b/xrdp/xrdp_mm.c
index f332776..a12b534 100644
--- a/xrdp/xrdp_mm.c
+++ b/xrdp/xrdp_mm.c
@@ -661,7 +661,7 @@ xrdp_mm_process_login_response(struct xrdp_mm* self, struct stream* s)
         {
           /* unix socket */
           self->chan_trans = trans_create(2, 8192, 8192);
-          g_snprintf(port, 255, "/tmp/xrdp_chansrv_socket_%d", 7200 + display);
+          g_snprintf(port, 255, "/var/run/xrdp/xrdp_chansrv_socket_%d", 7200 + display);
         }
         else
         {
