From: Dominik George <nik@naturalnet.de>
Subject: Use xorgxrdp instead of x11rdp
--- a/sesman/sesman.ini
+++ b/sesman/sesman.ini
@@ -54,12 +54,13 @@ LogLevel=DEBUG
 EnableSyslog=1
 SyslogLevel=DEBUG
 
-[X11rdp]
-param=X11rdp
-param=-bs
+[Xorg]
+param=Xorg
+param=-config
+param=xrdp/xorg.conf
+param=-noreset
 param=-nolisten
 param=tcp
-param=-uds
 
 [Xvnc]
 param=Xvnc
@@ -70,16 +71,6 @@ param=-localhost
 param=-dpi
 param=96
 
-[Xorg]
-param=Xorg
-param=-config
-param=xrdp/xorg.conf
-param=-noreset
-param=-nolisten
-param=tcp
-param=-logfile
-param=.xorgxrdp.%s.log
-
 [Chansrv]
 ; drive redirection, defaults to xrdp_client if not set
 FuseMountName=thinclient_drives
