From: Thorsten Glaser <tg@mirbsd.org>
Subject: config patches for Debian
 Add -retro option
 Use xorgxrdp instead of x11rdp
--- a/sesman/sesman.ini
+++ b/sesman/sesman.ini
@@ -61,12 +61,14 @@ LogLevel=DEBUG
 EnableSyslog=1
 SyslogLevel=DEBUG
 
-[X11rdp]
-param=X11rdp
-param=-bs
+[Xorg]
+param=Xorg
+param=-config
+param=xrdp/xorg.conf
+param=-noreset
 param=-nolisten
 param=tcp
-param=-uds
+param=-retro
 
 [Xvnc]
 param=Xvnc
@@ -77,16 +79,6 @@ param=-localhost
 param=-dpi
 param=96
 
-[Xorg]
-param=Xorg
-param=-config
-param=xrdp/xorg.conf
-param=-noreset
-param=-nolisten
-param=tcp
-param=-logfile
-param=.xorgxrdp.%s.log
-
 [Chansrv]
 ; drive redirection, defaults to xrdp_client if not set
 FuseMountName=thinclient_drives
--- a/xrdp/xrdp.ini
+++ b/xrdp/xrdp.ini
@@ -145,16 +145,6 @@ tcutils=true
 ; Session types
 ;
 
-[X11rdp]
-name=X11rdp
-lib=libxup.so
-username=ask
-password=ask
-ip=127.0.0.1
-port=-1
-xserverbpp=24
-code=10
-
 [Xorg]
 name=Xorg
 lib=libxup.so
