From: Thorsten Glaser <tg@mirbsd.org>
Subject: config patches for Debian
 Add -retro option
 Use xorgxrdp instead of x11rdp
--- a/sesman/sesman.ini
+++ b/sesman/sesman.ini
@@ -84,8 +84,7 @@ param=xrdp/xorg.conf
 param=-noreset
 param=-nolisten
 param=tcp
-param=-logfile
-param=/dev/null
+param=-retro
 
 [Chansrv]
 ; drive redirection, defaults to xrdp_client if not set
--- a/xrdp/xrdp.ini
+++ b/xrdp/xrdp.ini
@@ -145,16 +145,6 @@ tcutils=true
 ; Session types
 ;
 
-[X11rdp]
-name=X11rdp
-lib=libxup.so
-username=ask
-password=ask
-ip=127.0.0.1
-port=-1
-xserverbpp=24
-code=10
-
 [Xorg]
 name=Xorg
 lib=libxup.so
