From: jsorg71
Subject: Avoid 100% CPU load on ssl_tls_accept
Origin: https://github.com/neutrinolabs/xrdp/commit/a9eb21e6d73d94989dc0fa221824b0625b37b7aa.diff
Bug: https://github.com/neutrinolabs/xrdp/issues/954
Bug-Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=884453
--- a/common/ssl_calls.c
+++ b/common/ssl_calls.c
@@ -726,6 +726,15 @@ ssl_tls_accept(struct ssl_tls *self, int
              *     SSL_ERROR_WANT_READ
              *     SSL_ERROR_WANT_WRITE
              */
+            switch (SSL_get_error(self->ssl, connection_status))
+            {
+                case SSL_ERROR_WANT_READ:
+                    g_sck_can_recv(self->trans->sck, SSL_WANT_READ_WRITE_TIMEOUT);
+                    break;
+                case SSL_ERROR_WANT_WRITE:
+                    g_sck_can_send(self->trans->sck, SSL_WANT_READ_WRITE_TIMEOUT);
+                    break;
+            }
         }
         else
         {
