# DP: make FUSE mount root hidden from normal ls

--- a/sesman/chansrv/chansrv_fuse.c
+++ b/sesman/chansrv/chansrv_fuse.c
@@ -244,7 +244,7 @@ struct opendir_req
     struct fuse_file_info *fi;
 };
 
-static char g_fuse_mount_name[256] = "xrdp_client";
+static char g_fuse_mount_name[256] = ".xrdp_client";
 
 FIFO g_fifo_opendir;
 
@@ -456,7 +456,7 @@ xfuse_init(void)
 
     load_fuse_config();
 
-    /* define FUSE mount point to ~/xrdp_client, ~/thinclient_drives */
+    /* define FUSE mount point to ~/.xrdp_client, ~/.thinclient_drives */
     g_snprintf(g_fuse_root_path, 255, "%s/%s", g_getenv("HOME"), g_fuse_mount_name);
     g_snprintf(g_fuse_clipboard_path, 255, "%s/.clipboard", g_fuse_root_path);
 
--- a/sesman/sesman.ini
+++ b/sesman/sesman.ini
@@ -91,8 +91,8 @@ param6=tcp
 param7=-retro
 
 [Chansrv]
-# drive redirection, defaults to xrdp_client if not set
-FuseMountName=thinclient_drives
+# drive redirection, defaults to .xrdp_client if not set
+FuseMountName=.thinclient_drives
 
 [SessionVariables]
 PULSE_SCRIPT=/etc/xrdp/pulse/default.pa
