From: Tim Fischer
Subject: RFX fixes for large tile sets
Bug-Debian: http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=855387
Bug: https://github.com/neutrinolabs/xrdp/issues/524
--- a/xrdp/xrdp_encoder.c
+++ b/xrdp/xrdp_encoder.c
@@ -320,21 +320,16 @@ process_enc_rfx(struct xrdp_encoder *sel
     mutex = self->mutex;
     event_processed = self->xrdp_encoder_event_processed;
 
-    if ((enc->num_crects > 512) || (enc->num_drects > 512))
-    {
-        return 0;
-    }
-
     out_data_bytes = 16 * 1024 * 1024;
-    index = 256 + sizeof(struct rfx_tile) * 512 +
-                  sizeof(struct rfx_rect) * 512;
+    index = 256 + sizeof(struct rfx_tile) * enc->num_crects +
+                  sizeof(struct rfx_rect) * enc->num_drects;
     out_data = (char *) g_malloc(out_data_bytes + index, 0);
     if (out_data == 0)
     {
         return 0;
     }
     tiles = (struct rfx_tile *) (out_data + out_data_bytes + 256);
-    rfxrects = (struct rfx_rect *) (tiles + 512);
+    rfxrects = (struct rfx_rect *) (tiles + enc->num_crects);
 
     count = enc->num_crects;
     for (index = 0; index < count; index++)
