From: YOSHINO Yoshihito <yy.y.ja.jp@gmail.com>
Subject: xrdp: fails to detect some Japanese keyboard
Bug-Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=854847
Bug: https://github.com/neutrinolabs/xrdp/issues/663
--- a/xrdp/xrdp_keyboard.ini
+++ b/xrdp/xrdp_keyboard.ini
@@ -62,6 +62,9 @@ rdp_layout_de=0x00000407
 rdp_layout_fr=0x0000040C
 rdp_layout_it=0x00000410
 rdp_layout_jp=0x00000411
+rdp_layout_jp=0xe0010411
+rdp_layout_jp=0xe0200411
+rdp_layout_jp=0xe0210411
 rdp_layout_kr=0x00000412
 rdp_layout_ru=0x00000419
 rdp_layout_se=0x0000041D
