From: Thorsten Glaser <tg@mirbsd.org>
Subject: Recognise GNU/kFreeBSD as FreeBSD variant

--- a/common/os_calls.c
+++ b/common/os_calls.c
@@ -87,8 +87,8 @@ extern char **environ;
 #endif
 
 /* sys/ucred.h needs to be included to use struct xucred
- * in FreeBSD and OS X. No need for other BSDs  */
-#if defined(__FreeBSD__) || defined(__APPLE__)
+ * in FreeBSD and OS X. No need for other BSDs except GNU/kFreeBSD */
+#if defined(__FreeBSD__) || defined(__APPLE__) || defined(__FreeBSD_kernel__)
 #include <sys/ucred.h>
 #endif
 
--- a/librfxcodec/acinclude.m4
+++ b/librfxcodec/acinclude.m4
@@ -40,7 +40,7 @@ case "$host_os" in
         ;;
     esac
   ;;
-  freebsd* | netbsd* | openbsd*)
+  kfreebsd* | freebsd* | netbsd* | openbsd*)
     if echo __ELF__ | $CC -E - | grep __ELF__ > /dev/null; then
       objfmt='BSD-a.out'
     else
--- a/sesman/session.c
+++ b/sesman/session.c
@@ -503,7 +503,7 @@ session_start_fork(tbus data, tui8 type,
         g_sprintf(geometry, "%dx%d", s->width, s->height);
         g_sprintf(depth, "%d", s->bpp);
         g_sprintf(screen, ":%d", display);
-#ifdef __FreeBSD__
+#if defined(__FreeBSD__) || defined(__FreeBSD_kernel__)
         /*
          * FreeBSD bug
          * ports/157282: effective login name is not set by xrdp-sesman
--- a/xorgxrdp/acinclude.m4
+++ b/xorgxrdp/acinclude.m4
@@ -40,7 +40,7 @@ case "$host_os" in
         ;;
     esac
   ;;
-  freebsd* | netbsd* | openbsd*)
+  kfreebsd* | freebsd* | netbsd* | openbsd*)
     if echo __ELF__ | $CC -E - | grep __ELF__ > /dev/null; then
       objfmt='BSD-a.out'
     else
