From: Thorsten Glaser <tg@mirbsd.org>
Subject: Enable Large File Support on (at least) 32-bit Linux

--- a/common/pixman-region.c
+++ b/common/pixman-region.c
@@ -63,6 +63,8 @@
  * PERFORMANCE OF THIS SOFTWARE.
  */
 
+#include "arch.h"
+
 #define PIXREGION_NIL(reg) ((reg)->data && !(reg)->data->numRects)
 /* not a region */
 #define PIXREGION_NAR(reg)      ((reg)->data == pixman_broken_data)
--- a/configure.ac
+++ b/configure.ac
@@ -6,6 +6,8 @@ AC_CONFIG_HEADERS(config_ac.h:config_ac-
 AM_INIT_AUTOMAKE([1.7.2 foreign])
 AC_CONFIG_MACRO_DIR([m4])
 AC_PROG_CC
+AC_SYS_LARGEFILE
+AC_FUNC_FSEEKO
 AC_C_CONST
 AC_PROG_LIBTOOL
 PKG_PROG_PKG_CONFIG
--- a/genkeymap/evdev-map.c
+++ b/genkeymap/evdev-map.c
@@ -22,6 +22,8 @@
  * xfree86(base)->evdev keycode mapping
  */
 
+#include "arch.h"
+
 int xfree86_to_evdev[137-8+1] = {
     /* MDSW */ 203,
     /* ESC */ 9,
--- a/genkeymap/genkeymap.c
+++ b/genkeymap/genkeymap.c
@@ -39,6 +39,7 @@
 #include <X11/Xutil.h>
 #include <X11/XKBlib.h>
 #include <locale.h>
+#include "arch.h"
 
 extern int xfree86_to_evdev[137-8];
 
--- a/libpainter/configure.ac
+++ b/libpainter/configure.ac
@@ -6,6 +6,8 @@ AC_CONFIG_HEADERS(config_ac.h:config_ac-
 AM_INIT_AUTOMAKE([1.6 foreign])
 AC_CONFIG_MACRO_DIR([m4])
 AC_PROG_CC
+AC_SYS_LARGEFILE
+AC_FUNC_FSEEKO
 AC_C_CONST
 AC_PROG_LIBTOOL
 PKG_INSTALLDIR
--- a/librfxcodec/configure.ac
+++ b/librfxcodec/configure.ac
@@ -7,6 +7,8 @@ AM_INIT_AUTOMAKE([1.6 foreign])
 AC_CONFIG_MACRO_DIR([m4])
 m4_ifdef([AM_SILENT_RULES], [AM_SILENT_RULES])
 AC_PROG_CC
+AC_SYS_LARGEFILE
+AC_FUNC_FSEEKO
 AC_C_CONST
 AC_PROG_LIBTOOL
 PKG_INSTALLDIR
--- a/sesman/chansrv/pcsc/xrdp_pcsc.c
+++ b/sesman/chansrv/pcsc/xrdp_pcsc.c
@@ -10,6 +10,7 @@
 #include <sys/socket.h>
 #include <sys/un.h>
 #include <sys/stat.h>
+#include "arch.h"
 
 #define PCSC_API
 
--- a/sesman/tools/dis.c
+++ b/sesman/tools/dis.c
@@ -22,6 +22,7 @@
 #include <unistd.h>
 #include <sys/socket.h>
 #include <sys/un.h>
+#include "arch.h"
 
 int main(int argc, char **argv)
 {
--- a/sesman/tools/xcon.c
+++ b/sesman/tools/xcon.c
@@ -22,6 +22,7 @@
 #include <unistd.h>
 #include <sys/types.h>
 #include <X11/Xlib.h>
+#include "arch.h"
 
 Display *g_display = 0;
 int g_x_socket = 0;
--- a/xorgxrdp/configure.ac
+++ b/xorgxrdp/configure.ac
@@ -6,6 +6,8 @@ AC_CONFIG_HEADERS(config_ac.h:config_ac-
 AM_INIT_AUTOMAKE([1.11 foreign parallel-tests])
 m4_ifdef([AM_SILENT_RULES], [AM_SILENT_RULES([yes])])
 AC_PROG_CC
+AC_SYS_LARGEFILE
+AC_FUNC_FSEEKO
 AC_C_CONST
 AC_PROG_LIBTOOL
 
--- a/xorgxrdp/module/rdp.h
+++ b/xorgxrdp/module/rdp.h
@@ -30,6 +30,7 @@ CONNECTION WITH THE SOFTWARE OR THE USE
 #include <gcstruct.h>
 #include <mipointer.h>
 #include <randrstr.h>
+#include <config_ac.h>
 
 #include "rdpPri.h"
 
--- a/xorgxrdp/module/rdpReg.c
+++ b/xorgxrdp/module/rdpReg.c
@@ -33,6 +33,7 @@ to deal with regions changing in xorg ve
 #include <xf86.h>
 #include <xf86_OSproc.h>
 
+#include <config_ac.h>
 #include "rdpReg.h"
 
 /*
--- a/xrdpvr/xrdpvr.c
+++ b/xrdpvr/xrdpvr.c
@@ -21,6 +21,7 @@
  *
  */
 
+#include "arch.h"
 #include "xrdpvr.h"
 #include "xrdpvr_internal.h"
 
