# DP: misc. fixes, should go upstream

--- a/docs/man/xrdp-dis.1
+++ b/docs/man/xrdp-dis.1
@@ -1,4 +1,4 @@
-.TH "xrdp-dis" "8" "0.7.0" "xrdp team"
+.TH "xrdp-dis" "1" "0.7.0" "xrdp team"
 .SH NAME
 xrdp\-dis \- xrdp disconnect utility
 
@@ -7,7 +7,7 @@ xrdp\-dis \- xrdp disconnect utility
 
 .SH DESCRIPTION
 .PP
-\fBxrdp\-dix\fP is run with no parameters to disconnect your xrdp session.
+\fBxrdp\-dis\fP is run with no parameters to disconnect your xrdp session.
 
 .SH ENVIRONMENT
 .TP
--- a/librfxcodec/configure.ac
+++ b/librfxcodec/configure.ac
@@ -9,7 +9,7 @@ AC_PROG_CC
 AC_C_CONST
 AC_PROG_LIBTOOL
 
-AM_CONDITIONAL(GOT_PREFIX, test "x${prefix}" != "xNONE"])
+AM_CONDITIONAL(GOT_PREFIX, [test "x${prefix}" != "xNONE" && test "x${prefix}" != "x/usr"])
 
 if test "x${prefix}" = "xNONE" ; then
 sysconfdir="/etc";
--- a/sesman/chansrv/chansrv_fuse.c
+++ b/sesman/chansrv/chansrv_fuse.c
@@ -58,6 +58,9 @@ char g_fuse_clipboard_path[256] = ""; /*
 
 #include "arch.h"
 #include "chansrv_fuse.h"
+#include "chansrv.h"
+#include "devredir.h"
+#include "file.h"
 
 /* dummy calls when XRDP_FUSE is not defined */
 int xfuse_init(void)            { return 0; }
@@ -104,6 +107,9 @@ void xfuse_devredir_cb_file_close(void *
 #include "chansrv_fuse.h"
 #include "list.h"
 #include "fifo.h"
+#include "chansrv.h"
+#include "devredir.h"
+#include "file.h"
 
 #ifndef EREMOTEIO
 #define EREMOTEIO EIO
@@ -122,6 +128,7 @@ void xfuse_devredir_cb_file_close(void *
 #define LOG_ERROR   0
 #define LOG_INFO    1
 #define LOG_DEBUG   2
+#undef LOG_LEVEL
 #define LOG_LEVEL   LOG_ERROR
 
 #define log_error(_params...)                           \
--- a/sesman/chansrv/devredir.h
+++ b/sesman/chansrv/devredir.h
@@ -101,6 +101,8 @@ int dev_redir_file_open(void *fusep, tui
 
 int devredir_file_close(void *fusep, tui32 device_id, tui32 file_id);
 
+int APP_CC devredir_rmdir_or_file(void *fusep, tui32 device_id, char *path, int mode);
+
 int devredir_file_read(void *fusep, tui32 device_id, tui32 FileId,
                         tui32 Length, tui64 Offset);
 
--- a/sesman/session.c
+++ b/sesman/session.c
@@ -609,7 +609,7 @@ session_start_fork(int width, int height
                 else
                 {
                     log_message(LOG_LEVEL_ERROR, "another Xserver might "
-                                "already be active on display %d - see log", display);
+                                "already be active on display %d - see log; possibly comment out the two param lines -logfile /dev/null in /etc/xrdp/sesman.ini to debug", display);
                 }
 
                 log_message(LOG_LEVEL_DEBUG, "aborting connection...");
--- a/xorgxrdp/configure.ac
+++ b/xorgxrdp/configure.ac
@@ -9,7 +9,7 @@ AC_PROG_CC
 AC_C_CONST
 AC_PROG_LIBTOOL
 
-AM_CONDITIONAL(GOT_PREFIX, test "x${prefix}" != "xNONE"])
+AM_CONDITIONAL(GOT_PREFIX, [test "x${prefix}" != "xNONE" && test "x${prefix}" != "x/usr"])
 
 AC_CHECK_HEADER([xorg/xorg-server.h], [],
   [AC_MSG_ERROR([please install xserver-xorg-dev or xorg-x11-server-sdk])])
--- a/xorgxrdp/xrdpkeyb/rdpKeyboard.c
+++ b/xorgxrdp/xrdpkeyb/rdpKeyboard.c
@@ -525,7 +525,7 @@ rdpInputKeyboard(rdpPtr dev, int msg, lo
 }
 
 /******************************************************************************/
-void
+int
 rdpkeybDeviceInit(DeviceIntPtr pDevice, KeySymsPtr pKeySyms, CARD8 *pModMap)
 {
     int i;
@@ -557,7 +557,7 @@ rdpkeybDeviceInit(DeviceIntPtr pDevice,
     if (pKeySyms->map == 0)
     {
         LLOGLN(0, ("rdpkeybDeviceInit: malloc failed"));
-        exit(1);
+        return 1;
     }
     else
     {
@@ -573,6 +573,8 @@ rdpkeybDeviceInit(DeviceIntPtr pDevice,
     {
         pKeySyms->map[i] = g_kbdMap[i];
     }
+
+    return 0;
 }
 
 /******************************************************************************/
@@ -668,7 +670,8 @@ rdpkeybControl(DeviceIntPtr device, int
     switch (what)
     {
         case DEVICE_INIT:
-            rdpkeybDeviceInit(device, &keySyms, modMap);
+            if (rdpkeybDeviceInit(device, &keySyms, modMap))
+                return BadAlloc;
             memset(&set, 0, sizeof(set));
             set.rules = "base";
             set.model = "pc104";
--- a/xrdpapi/xrdpapi.c
+++ b/xrdpapi/xrdpapi.c
@@ -36,6 +36,7 @@
 #include <sys/un.h>
 
 #include "xrdpapi.h"
+#include "os_calls.h"
 
 struct wts_obj
 {
