# DP: re-enable non-executable stack

--- a/librfxcodec/src/amd64/cpuid_amd64.asm
+++ b/librfxcodec/src/amd64/cpuid_amd64.asm
@@ -39,3 +39,4 @@ PROC cpuid_amd64
     ret;
     align 16
 
+section .note.GNU-stack noalloc noexec nowrite progbits
--- a/librfxcodec/src/x86/cpuid_x86.asm
+++ b/librfxcodec/src/x86/cpuid_x86.asm
@@ -37,3 +37,4 @@ PROC cpuid_x86
     ret;
     align 16
 
+section .note.GNU-stack noalloc noexec nowrite progbits
--- a/xorgxrdp/module/amd64/a8r8g8b8_to_a8b8g8r8_box_amd64_sse2.asm
+++ b/xorgxrdp/module/amd64/a8r8g8b8_to_a8b8g8r8_box_amd64_sse2.asm
@@ -187,3 +187,4 @@ done_loop_x:
     ret
     align 16
 
+section .note.GNU-stack noalloc noexec nowrite progbits
--- a/xorgxrdp/module/amd64/a8r8g8b8_to_nv12_box_amd64_sse2.asm
+++ b/xorgxrdp/module/amd64/a8r8g8b8_to_nv12_box_amd64_sse2.asm
@@ -322,3 +322,4 @@ loop1:
     ret
     align 16
 
+section .note.GNU-stack noalloc noexec nowrite progbits
--- a/xorgxrdp/module/amd64/cpuid_amd64.asm
+++ b/xorgxrdp/module/amd64/cpuid_amd64.asm
@@ -39,3 +39,4 @@ PROC cpuid_amd64
     ret
     align 16
 
+section .note.GNU-stack noalloc noexec nowrite progbits
--- a/xorgxrdp/module/amd64/i420_to_rgb32_amd64_sse2.asm
+++ b/xorgxrdp/module/amd64/i420_to_rgb32_amd64_sse2.asm
@@ -242,3 +242,4 @@ loop_x:
     align 16
 
 
+section .note.GNU-stack noalloc noexec nowrite progbits
--- a/xorgxrdp/module/amd64/uyvy_to_rgb32_amd64_sse2.asm
+++ b/xorgxrdp/module/amd64/uyvy_to_rgb32_amd64_sse2.asm
@@ -152,3 +152,4 @@ loop1:
     ret
     align 16
 
+section .note.GNU-stack noalloc noexec nowrite progbits
--- a/xorgxrdp/module/amd64/yuy2_to_rgb32_amd64_sse2.asm
+++ b/xorgxrdp/module/amd64/yuy2_to_rgb32_amd64_sse2.asm
@@ -152,3 +152,4 @@ loop1:
     ret
     align 16
 
+section .note.GNU-stack noalloc noexec nowrite progbits
--- a/xorgxrdp/module/amd64/yv12_to_rgb32_amd64_sse2.asm
+++ b/xorgxrdp/module/amd64/yv12_to_rgb32_amd64_sse2.asm
@@ -246,3 +246,4 @@ loop_x:
     align 16
 
 
+section .note.GNU-stack noalloc noexec nowrite progbits
--- a/xorgxrdp/module/x86/a8r8g8b8_to_a8b8g8r8_box_x86_sse2.asm
+++ b/xorgxrdp/module/x86/a8r8g8b8_to_a8b8g8r8_box_x86_sse2.asm
@@ -172,3 +172,4 @@ done_loop_x:
     ret
     align 16
 
+section .note.GNU-stack noalloc noexec nowrite progbits
--- a/xorgxrdp/module/x86/a8r8g8b8_to_nv12_box_x86_sse2.asm
+++ b/xorgxrdp/module/x86/a8r8g8b8_to_nv12_box_x86_sse2.asm
@@ -316,3 +316,4 @@ loop1:
     ret
     align 16
 
+section .note.GNU-stack noalloc noexec nowrite progbits
--- a/xorgxrdp/module/x86/cpuid_x86.asm
+++ b/xorgxrdp/module/x86/cpuid_x86.asm
@@ -37,3 +37,4 @@ PROC cpuid_x86
     ret
     align 16
 
+section .note.GNU-stack noalloc noexec nowrite progbits
--- a/xorgxrdp/module/x86/i420_to_rgb32_x86_sse2.asm
+++ b/xorgxrdp/module/x86/i420_to_rgb32_x86_sse2.asm
@@ -241,3 +241,4 @@ loop_x:
     align 16
 
 
+section .note.GNU-stack noalloc noexec nowrite progbits
--- a/xorgxrdp/module/x86/uyvy_to_rgb32_x86_sse2.asm
+++ b/xorgxrdp/module/x86/uyvy_to_rgb32_x86_sse2.asm
@@ -156,3 +156,4 @@ loop1:
     ret
     align 16
 
+section .note.GNU-stack noalloc noexec nowrite progbits
--- a/xorgxrdp/module/x86/yuy2_to_rgb32_x86_sse2.asm
+++ b/xorgxrdp/module/x86/yuy2_to_rgb32_x86_sse2.asm
@@ -156,3 +156,4 @@ loop1:
     ret
     align 16
 
+section .note.GNU-stack noalloc noexec nowrite progbits
--- a/xorgxrdp/module/x86/yv12_to_rgb32_x86_sse2.asm
+++ b/xorgxrdp/module/x86/yv12_to_rgb32_x86_sse2.asm
@@ -241,3 +241,4 @@ loop_x:
     align 16
 
 
+section .note.GNU-stack noalloc noexec nowrite progbits
