From: Thorsten Glaser <tg@mirbsd.org>
Subject: don’t spit on the console when starting
--- a/common/log.c
+++ b/common/log.c
@@ -383,11 +383,6 @@ internal_config_read_logging(int file, s
     /* try to create path if not exist */
     g_create_path(lc->log_file);
 
-    g_printf("logging configuration:\r\n");
-    g_printf("\tLogFile:       %s\r\n", lc->log_file);
-    g_printf("\tLogLevel:      %i\r\n", lc->log_level);
-    g_printf("\tEnableSyslog:  %i\r\n", lc->enable_syslog);
-    g_printf("\tSyslogLevel:   %i\r\n", lc->syslog_level);
     return LOG_STARTUP_OK;
 }
 
--- a/sesman/sesman.c
+++ b/sesman/sesman.c
@@ -149,6 +149,12 @@ main(int argc, char **argv)
 
     if (1 == argc)
     {
+	/* redirect stdout to logfile early, due to debug msgs */
+	if (g_file_exist("/var/log/xrdp-sesman.log")) {
+		g_file_close(1);
+		g_file_open("/var/log/xrdp-sesman.log");
+	}
+
         /* no options on command line. normal startup */
         g_printf("starting sesman...\n");
         daemon = 1;
--- a/xrdp/xrdp.c
+++ b/xrdp/xrdp.c
@@ -544,7 +544,6 @@ main(int argc, char **argv)
 
         if (0 != pid)
         {
-            g_writeln("process %d started ok", pid);
             /* exit, this is the main process */
             g_deinit();
             g_exit(0);
