Subject: Fixes to systemd service files
From: Dominik George <nik@naturalnet.de>
Forwarded: https://github.com/neutrinolabs/xrdp/pull/646
--- a/instfiles/xrdp-sesman.service
+++ b/instfiles/xrdp-sesman.service
@@ -1,12 +1,13 @@
 [Unit]
 Description=xrdp session manager
-After=syslog.target network.target
+Documentation=man:xrdp-sesman(8) man:sesman.ini(5)
+After=network.target
 StopWhenUnneeded=true
-BindTo=xrdp.service
+BindsTo=xrdp.service
 
 [Service]
 Type=forking
-PIDFile=/var/run/xrdp-sesman.pid
+PIDFile=/run/xrdp/xrdp-sesman.pid
 EnvironmentFile=-/etc/sysconfig/xrdp
 EnvironmentFile=-/etc/default/xrdp
 ExecStart=/usr/sbin/xrdp-sesman $SESMAN_OPTIONS
--- a/instfiles/xrdp.service
+++ b/instfiles/xrdp.service
@@ -1,13 +1,16 @@
 [Unit]
 Description=xrdp daemon
+Documentation=man:xrdp(8) man:xrdp.ini(5)
 Requires=xrdp-sesman.service
-After=syslog.target network.target xrdp-sesman.service
+After=network.target xrdp-sesman.service
 
 [Service]
 Type=forking
-PIDFile=/var/run/xrdp.pid
+PIDFile=/run/xrdp/xrdp.pid
 EnvironmentFile=-/etc/sysconfig/xrdp
 EnvironmentFile=-/etc/default/xrdp
+User=xrdp
+Group=xrdp
 PermissionsStartOnly=true
 ExecStartPre=/bin/sh /usr/share/xrdp/socksetup
 ExecStart=/usr/sbin/xrdp $XRDP_OPTIONS
