Subject: Fixes to systemd service files
From: Dominik George <nik@naturalnet.de>
Forwarded: https://github.com/neutrinolabs/xrdp/pull/646

--- a/instfiles/xrdp-sesman.service.in
+++ b/instfiles/xrdp-sesman.service.in
@@ -7,6 +7,7 @@
 
 [Service]
 Type=exec
+RuntimeDirectory=xrdp
 EnvironmentFile=-@sysconfdir@/sysconfig/xrdp
 EnvironmentFile=-@sysconfdir@/default/xrdp
 ExecStart=@sbindir@/xrdp-sesman $SESMAN_OPTIONS --nodaemon
--- a/instfiles/xrdp.service.in
+++ b/instfiles/xrdp.service.in
@@ -6,8 +6,13 @@
 
 [Service]
 Type=exec
+RuntimeDirectory=xrdp
 EnvironmentFile=-@sysconfdir@/sysconfig/xrdp
 EnvironmentFile=-@sysconfdir@/default/xrdp
+User=xrdp
+Group=xrdp
+PermissionsStartOnly=true
+ExecStartPre=/bin/sh /usr/share/xrdp/socksetup
 ExecStart=@sbindir@/xrdp $XRDP_OPTIONS --nodaemon
 
 [Install]
