Subject: Fixes to systemd service files
From: Dominik George <nik@naturalnet.de>
Forwarded: https://github.com/neutrinolabs/xrdp/pull/646
--- a/instfiles/xrdp-sesman.service
+++ b/instfiles/xrdp-sesman.service
@@ -7,7 +7,8 @@ BindsTo=xrdp.service
 
 [Service]
 Type=forking
-PIDFile=/var/run/xrdp-sesman.pid
+PIDFile=/run/xrdp/xrdp-sesman.pid
+RuntimeDirectory=xrdp
 EnvironmentFile=-/etc/sysconfig/xrdp
 EnvironmentFile=-/etc/default/xrdp
 ExecStart=/usr/sbin/xrdp-sesman $SESMAN_OPTIONS
--- a/instfiles/xrdp.service
+++ b/instfiles/xrdp.service
@@ -6,9 +6,12 @@ After=network.target xrdp-sesman.service
 
 [Service]
 Type=forking
-PIDFile=/var/run/xrdp.pid
+PIDFile=/run/xrdp/xrdp.pid
+RuntimeDirectory=xrdp
 EnvironmentFile=-/etc/sysconfig/xrdp
 EnvironmentFile=-/etc/default/xrdp
+User=xrdp
+Group=xrdp
 PermissionsStartOnly=true
 ExecStartPre=/bin/sh /usr/share/xrdp/socksetup
 ExecStart=/usr/sbin/xrdp $XRDP_OPTIONS
