--- a/xrdp/xrdp_mm.c	2022-09-30 18:11:21.920146026 +0200
+++ b/xrdp/xrdp_mm.c	2022-09-30 18:54:17.562336406 +0200
@@ -1216,18 +1216,14 @@
                 best_pro_index = index;
                 break;
             case XR_RDPGFX_CAPVERSION_104:
+            case XR_RDPGFX_CAPVERSION_105:
+            case XR_RDPGFX_CAPVERSION_106:
                 if (!(flags & XR_RDPGFX_CAPS_FLAG_AVC_DISABLED))
                 {
                     best_h264_index = index;
                 }
                 best_pro_index = index;
                 break;
-            case XR_RDPGFX_CAPVERSION_105:
-                best_pro_index = index;
-                break;
-            case XR_RDPGFX_CAPVERSION_106:
-                best_pro_index = index;
-                break;
         }
     }
     if (best_pro_index >= 0)
--- a/xrdp/xrdp_encoder.c	2022-09-30 18:11:21.916146016 +0200
+++ b/xrdp/xrdp_encoder.c	2022-09-30 18:53:04.405824688 +0200
@@ -629,6 +629,7 @@
     g_memset(s, 0, sizeof(struct stream));
     ls.data = out_data + 256;
     ls.p = ls.data;
+    ls.size = out_data_bytes + index - 256;
 
     if (self->gfx)
     {
